### PR TITLE
Codal listener

### PR DESCRIPTION
**Pull request by tballmsft**
Changes to enable multiple background handlers for an event (simulator). Doesn't affect behavior for user handlers (last handler added wins).  Test program (requires minor pxt-microbit changes) at https://makecode.microbit.org/_M499UaitjCT3

### CODE REVIEW

**Comment by tballmsft** [comment]
@riknoll - can you please take a look?

File: pxtsim/runtime.ts [hunk 1]
```diff
@@ -223,7 +223,7 @@ namespace pxsim {
                     aws.forEach(aw => aw());
                 }
             }
-            if (!this.handler || this.events.length > this.max) return;
+            if (this.handlers == [] || this.events.length > this.max) return;
```
Comment (mmoskal) [inline pxtsim/runtime.ts]: should be `this.handlers.length == 0`; `==` is probably ref-equality

File: pxtsim/runtime.ts [hunk 2]
```diff
@@ -245,21 +246,30 @@ namespace pxsim {
                         this.lock = false;
```
Comment (mmoskal) [inline pxtsim/runtime.ts]: You're releasing the lock multiple times it seems

File: pxtsim/runtime.ts [hunk 3]
```diff
@@ -245,21 +246,30 @@ namespace pxsim {
...
-                pxtcore.decr(this._handler);
-            }
+        addHandler(a: RefAction) {
+            this._handlers.push(a);
+            pxtcore.incr(a)
+        }
 
-            this._handler = a;
+        setHandler(a: RefAction) {
+            this._handlers.forEach(old => pxtcore.decr(old))
```
Comment (mmoskal) [inline pxtsim/runtime.ts]: This seems to remove all old handlers, not only the foreground one

File: pxtsim/simlib.ts [hunk 4]
```diff
@@ -37,34 +43,56 @@ namespace pxsim {
...
 
         queue(id: number | string, evid: number | string, value: T = null) {
             // special handling for notify one
             const notifyOne = this.notifyID && this.notifyOneID && id == this.notifyOneID;
             if (notifyOne)
                 id = this.notifyID;
-
-            // grab queue and handle
+            // first, background
+            this.backgroundHandlerFlag = true;
```
Comment (mmoskal) [inline pxtsim/simlib.ts]: This flag seems to be used for two different purposes - for setting handlers and for running them. Two flags would be better, especially when handlers register new handlers.

File: pxtsim/runtime.ts (hunk 1 above)
Comment (tballmsft) [reply pxtsim/runtime.ts]: thanks.

File: pxtsim/runtime.ts (hunk 3 above)
Comment (tballmsft) [reply pxtsim/runtime.ts]: Good catch.

Comment (tballmsft) [reply pxtsim/runtime.ts]: but setHandler only called for foreground.

Comment (mmoskal) [reply pxtsim/runtime.ts]: Oh, I think I misunderstood.

File: pxtsim/simlib.ts [hunk 5]
```diff
@@ -37,34 +43,56 @@ namespace pxsim {
...
+            else
+                q.setHandler(handler);
+            this.backgroundHandlerFlag = false;
+        }
+
+        // only for background handlers
+        remove(handler: RefAction) {
+            this.backgroundHandlerFlag = false;
+            Object.keys(this.queues).forEach((k: string) => {
+                if (k.startsWith("back"))
```
Comment (mmoskal) [inline pxtsim/simlib.ts]: should be `"back:"` ?

**Comment by pelikhan** [comment]
Any docs?

**Comment by tballmsft** [comment]
Not yet, this is just plumbing. No user facing APIs surfaced here. That will come in common-packages.

**Pull request merged**